lib/triton_kernel_call.cc

```diff
@@ -0,0 +1,124 @@
+/* Copyright 2022 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include "triton_kernel_call.h"
+
+#include <iostream>
+#include <string>
+
+#include <pybind11/pybind11.h>
+#include "cuda.h"
+
+namespace py = pybind11;
+
+void TritonExecutable::launch(CUstream stream, void** buffers) {
+  CUdevice dev;
+  CUcontext ctx;
+  // Set the current context to the stream context so we can query the stream
+  // device
+  cuStreamGetCtx(stream, &ctx);
+  cuCtxSetCurrent(ctx);
+  /// Only load the kernel if it hasn't already been loaded for this device
+  cuCtxGetDevice(&dev);
+  if (!is_loaded(dev)) {
```

Comment:
I think you need a lock around loading. Else you can get a race.

Reply:
Done.